LIBITD-2553 Added the single (detail) group view

aatishgupta25 opened the pull request:
Updated the urls.py and views.py to include the single group view and created a new template ui/single_group_view.html

https://umd-dit.atlassian.net/browse/LIBITD-2553

Commit:
LIBITD-2553 Added the single (detail) group view

Updated the urls.py and views.py to include the single
group view and created a new template ui/single_group_view.html

https://umd-dit.atlassian.net/browse/LIBITD-2553

peichman-umd commented on src/ipmanager/ui/templates/ui/single_group_view.html:
```diff
@@ -0,0 +1,76 @@
+{% block content %}
+<h1> {{group.key}} </h1>
+
+<h4 style="margin-bottom: 0.5rem;">Group Information</h4>
```
This (and the other `<h4>` tags) should be `<h2>` instead. For accessibility reasons, do not skip levels in the `h1`-`h6` tags:

> A common navigation technique for users of screen reading software is to quickly jump from heading to heading in order to determine the content of the page. Because of this, it is important to not skip one or more heading levels. Doing so may create confusion, as the person navigating this way may be left wondering where the missing heading is.

https://developer.mozilla.org/en-US/docs/Web/HTML/Element/Heading_Elements#accessibility

peichman-umd commented on src/ipmanager/ui/templates/ui/single_group_view.html:
```diff
@@ -0,0 +1,76 @@
+</div>
+<br>
+
+<h4 style="margin-bottom: 0.5rem;">IP Ranges</h4>
```
Change to `<h2>` as noted above.

peichman-umd commented on src/ipmanager/ui/templates/ui/single_group_view.html:
```diff
@@ -0,0 +1,76 @@
+
+<br>
+
+<h4 style="margin-bottom: 0.5rem;">Child Groups</h4>
```
Change to `<h2>` as noted above.

peichman-umd commented on src/ipmanager/ui/templates/ui/single_group_view.html:
```diff
@@ -0,0 +1,76 @@
+<h1> {{group.key}} </h1>
+
+<h4 style="margin-bottom: 0.5rem;">Group Information</h4>
+<div style="width: 80vw; background-color: rgb(196, 203, 201);">
```
For now, I would suggest leaving off any inline style attributes. In my experience, it is better to do the styling as a second pass over the application after all the views and forms are in place. Having everything in place sometimes exposes patterns that are harder to see in the isolation of a single view.

peichman-umd commented on src/ipmanager/ui/views.py:
```diff
@@ -15,4 +16,22 @@ class GroupListView(ListView):
+
+    current_group = self.object
+    
+    context['ip_ranges'] = IPRange.objects.filter(group=current_group)
```
An alternate Python idiom to use here is the `update()` method of dictionaries (https://docs.python.org/3/library/stdtypes.html#dict.update):

```python
context.update(
    ip_ranges=IPRange.objects.filter(group=current_group),
    included_groups=Relation.objects.filter(subject=current_group, relation=Relation.RelationType.INCLUSION),
    excluded_groups=Relation.objects.filter(subject=current_group, relation=Relation.RelationType.EXCLUSION),
)
```

Another way to approach this would be to add additional [properties](https://docs.python.org/3/library/functions.html#property) to the `Group` model class to encapsulate the `.objects.filter(...)` calls:

```python
# in models.py
class Group(Model):
    ...
    @property
    def ip_ranges(self):
        return IPRange.objects.filter(group=self)

    @property
    def included_groups(self):
        return Relation.objects.filter(subject=self, relation=Relation.RelationType.INCLUSION)

    @property
    def excluded_groups(self):
        return Relation.objects.filter(subject=self, relation=Relation.RelationType.EXCLUSION)
```
Then, in the template, you would have access to `group.ip_ranges`, `group.included_groups` and `group.excluded_groups`, with no need to update the context.

peichman-umd reviewed (changes requested):
Logic is good. I have some inline comments on the HTML and CSS, and some more advanced method encapsulation for the `Groups` model.

Commit:
LIBITD-2553 Edited Single Group View

Changed H4 tags to H2. Removed incline CSS.

https://umd-dit.atlassian.net/browse/LIBITD-2553
